### PR TITLE
article about pipeline mode

### PR DESCRIPTION
**Pull request by dlax**
@dvarrazzo, as discussed, this is a reuse of the [article on dalibo's blog](https://blog.dalibo.com/2022/09/19/psycopg-pipeline-mode.html), with (minor) changes in the second commit. I think we can change more things if appropriate, let me know.

### CODE REVIEW

File: content/articles/psycopg3-pipeline-mode/contents.lr [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
...ffffffe9&\xffffffa5\x1d\xffffffbd\x05\xfffffff3\x0b\xffffff9c\x1f\xffffffef\x00\xfffffffe\x05X :\x09C\x08\x12\x19\xfffffff2\x07\x1f\x06\xfffffffb\x03j\x00\xffffffe4\x0c\xffffff91\x10\xffffff94&l"\xffffffc3\x0e?\x04&\x0f+\x04\xffffffd1\x18q )\x13\x0d\x17\x10\x00\xffffffcf\x01\xffffffcb\x04\x03\x0b\xffffffe3\x01\xffffffe2\x16\xffffff8c\x1e\xfffffff5\x0f\xffffffee\x1b\xffffffcf\x01z&\x03\x02o\x10\xffffffd0\x1c\xffffffaf\x01\xfffffffb\x1f7\x05\xffffffcb\x0cL\x06r\x19&\x0a{\x15\x0a"\xffffffa1\x14\x05"N\x17\x0a\x11E\x04\x18\x1e\xffffffcd%\x0a\x1f\xfffffffd\x1b\xffffff87\x13\xffffff99\x0d\xffffff89\x0d\xffffff8e\x12\xffffff9a\x18g\x01\xfffffff8#\x1b\x12=#\xffffff97%\xffffff99\x1f\xffffffae$v#d#\xffffff8a\x15\xffffffed\x03G\x04P\x1e[\x0b`\x1d\x7f\x1e\xffffff9a&\xffffff9d&\xffffffe6\x08\xffffffcb\x1f.\x01M\x0c\xffffff82\x19\xfffffffe\x11F\x10\xffffffbd\x12#\x03\xffffffa5\x17\x1b\x18\xfffffff5\x18\xffffffd8"<\x0a\xffffff99\x17\xffffffba!1\x09\xffffffa2\x06\xffffffe0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' 8 '\x00\x02\xffffff8b\xffffff8f\xffffff82X\xffffffa6\xffffffe6' 1 '\x01' 1 0
```
Comment (dvarrazzo) [inline content/articles/psycopg3-pipeline-mode/contents.lr:252]: Maybe these strings can be shortened in the examples?

File: content/articles/psycopg3-pipeline-mode/contents.lr [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+As mentioned earlier, the entry point for the pipeline mode is the
+`pipeline()`_ method on ``Connection`` object which enters and exists
+pipeline mode. But what does this mean? Well, basically, this involves
+calling underlying `PQ{enter,exit}PipelineMode`_ functions.
+
+But this does not tell much about how things work in Psycopg.
+
+To actually understand how things work, we need to step back and read
+`libpq pipeline mode`_ documentation, in which section “Interleaving
+Result Processing and Query Dispatch” states: > The client application
```
Comment (dvarrazzo) [inline content/articles/psycopg3-pipeline-mode/contents.lr:285]: You can make a block quote in reST by indenting. So, untested:

```
Query Dispatch” states:

    The client application should generally maintain a queue of work remaining
    to be dispatched and a queue of work that has been dispatched but not yet
    had its
    ...
```

File: content/articles/psycopg3-pipeline-mode/contents.lr [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+should generally maintain a queue of work remaining > to be dispatched
+and a queue of work that has been dispatched but not yet > had its
+results processed. When the socket is writable it should dispatch > more
+work. When the socket is readable it should read results and process >
+them, matching them up to the next entry in its corresponding results
+queue.
+
+As often with PostgreSQL, everything is there. But this paragraph is
```
Comment (dvarrazzo) [inline content/articles/psycopg3-pipeline-mode/contents.lr:293]:
```suggestion
As often with PostgreSQL, everything is there, although this paragraph is
```

**Review by dvarrazzo — CHANGES_REQUESTED** [review]
Just some minor cleanups. Great article :)

File: content/articles/psycopg3-pipeline-mode/contents.lr [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+   CREATE TABLE t (x numeric, d timestamp, p boolean)
+
+and assume an application does a lot of queries like:
+
+.. code:: sql
+
+   INSERT INTO t (x, d, p) VALUES ($1, $2, $3)
+
+with distinct values, and with ``x`` possibly being a large integer
+(``n!, n<1000``). Maybe the application could make use of batch inserts
```
[the remark this answers is not in the document]
Comment (dlax) [reply content/articles/psycopg3-pipeline-mode/contents.lr]: Indeed, I have dropped this as it's not very meaningful after all.

**Comment by dlax** [comment]
The publication date is passed (last week); should I update it or leave it to you?

**Comment by dlax** [comment]
Seems we forgot about this[^1]; probably not worth any more or too late.

[^1]: I'm cleaning up old PRs

**Comment by dlax** [comment]
@dvarrazzo, as discussed, let's try again to republish this article to psycopg's blog; whenever you have time :)

File: content/articles/psycopg3-pipeline-mode/contents.lr [hunk 5]
```diff
@@ -0,0 +1,532 @@
+title: Pipeline mode in Psycopg
+---
+pub_date: 2022-10-14
```
Comment (dlax) [inline content/articles/psycopg3-pipeline-mode/contents.lr:3]: To be updated before publication.

Comment (dvarrazzo) [reply content/articles/psycopg3-pipeline-mode/contents.lr]: Set it to today

**Comment by dvarrazzo** [comment]
The article has been published, thank you very much :)

If it's ok I'll mention it on some social media post. Of course feel free to do the same!

**Comment by dlax** [comment]
> The article has been published, thank you very much :)

Great!

> If it's ok I'll mention it on some social media post.

Yes, of course, no problem.

**Pull request merged**